src/backend/foreign/foreign.c

```diff
@@ -334,6 +339,116 @@ GetUserMapping(Oid userid, Oid serverid)
 	return um;
 }
 
+List *
+GetForeignServerSegsByRelId(Oid relid)
+{
+	Form_pg_foreign_table_seg tablesegform;
+	Relation 	foreignTableRel;
+	ScanKeyData	ftkey;
+	SysScanDesc	ftscan;
+	HeapTuple	fttuple;
+	List	   *segList = NIL;
+
+
+	foreignTableRel = table_open(ForeignTableRelationSegId, AccessShareLock);
+
+	ScanKeyInit(&ftkey,
+				Anum_pg_foreign_table_seg_ftsrelid,
+				BTEqualStrategyNumber, F_OIDEQ,
+				ObjectIdGetDatum(relid));
+
+	ftscan = systable_beginscan(foreignTableRel, InvalidOid,
```

Comment:
If I don't misunderstand, the "GetForeignServerSegsByRelId" is called frequently. How about putting the system catalog table `ForeignTableRelationSegId` in the system cache instead of fetching rows from it by scanning the table every time?